pkgs/development/haskell-modules/generic-builder.nix

```diff
@@ -383,11 +383,20 @@ stdenv.mkDerivation ({
       mv tmp "$d"
     done
 
-    for d in $(grep '^dynamic-library-dirs:' "$packageConfDir"/* | cut -d' ' -f2- | tr ' ' '\n' | sort -u); do
-      for lib in "$d/"*.{dylib,so}; do
-        ln -s "$lib" "$dynamicLinksDir"
+    listDynamicLinks () {
+      for d in $(grep '^dynamic-library-dirs:' "$packageConfDir"/* | cut -d' ' -f2- | tr ' ' '\n' | sort -u); do
+        for lib in "$d/"*.{dylib,so} ; do
+          if [ -f "$lib" ] ; then
```

Comment:
When we use linux, lib includes not only actual filenames but also `*.dylib`.
This `[ -f "$lib" ]` is used to delete `*.dylib`.

Reply:
Mmm, I see -- because there are no files matching `somedir/*.dylib`?

Hmm, but this bit of code has always been this way -- in fact it's only since February in 019e86fde that it mentions `.so` at all, and before that it was only `.dylib`:
```diff
     for d in $(grep '^dynamic-library-dirs:' "$packageConfDir"/* | cut -d' ' -f2- | tr ' ' '\n' | sort -u); do
-      for lib in "$d/"*.dylib; do
+      for lib in "$d/"*.{dylib,so}; do
         ln -s "$lib" "$dynamicLinksDir"
```

This change should be a separate commit, in any event. Probably deserves some separate debugging and discussion, too. It sounds like an independent issue from the collision issue.